src/crypto/sodium.cpp

```diff
@@ -22,138 +22,190 @@
-    if (pubKey == NULL) {
-        // Something went wrong allocating the memory.
-        return(0);
-    }
-    unsigned char* privKey = (unsigned char*)malloc(crypto_sign_SECRETKEYBYTES);
-    if (privKey == NULL) {
-        // Something went wrong allocating the memory.
-        free(pubKey);
-        return(0);
-    }
-    int check = 0;
-    check = crypto_sign_seed_keypair(pubKey, privKey,
-                                     (const unsigned char*)this->keySeed);
-    if (check != 0) {
-        // Something went wrong deriving keys.
-        return(0);
-    }
-    unsigned long long bufferlen = 0;
-    unsigned char* sigbuffer = (unsigned char*)malloc(msglen + crypto_sign_BYTES);
-    if (sigbuffer == NULL) {
-        // Something went wrong allocating the memory.
-        free(pubKey);
-        free(privKey);
-        return(0);
-    }
-    check = crypto_sign(sigbuffer, &bufferlen, (const unsigned char*)msg, msglen,
-                        (const unsigned char*)privKey);
-    if (check != 0) {
-        // Something went wrong signing the message.
-        free(sigbuffer);
-        free(pubKey);
-        free(privKey);
-        return(0);
+int EdDSA::sign(const unsigned char* msg, const unsigned long long msglen,
+                unsigned char* sig)
+{
+    if (!sig || !msg)
+    {
+        return 0;
     }
-    free(sigbuffer);
-    free(pubKey);
-    free(privKey);
-    return(bufferlen);
+
+    int result = crypto_sign_detached(sig, NULL, msg, msglen, (const unsigned char*)privKey);
+
+    return (result == 0) ? (crypto_sign_BYTES + msglen) : 0;
 }
 
 
 // Verifies the signature of a message.
 int EdDSA::verify(const unsigned char* msg, unsigned long long msglen,
-                  const unsigned char* sig, const unsigned char* pubKey) {
-    unsigned char* msgbuffer = (unsigned char*)malloc(msglen + crypto_sign_BYTES);
-    if (msgbuffer == NULL) {
-        // Something went wrong allocating the memory.
-        return(0);
-    }
-    unsigned char* signedmsg = (unsigned char*)malloc(msglen + crypto_sign_BYTES);
-    if (signedmsg == NULL) {
-        // Something went wrong allocating the memory.
-        free(msgbuffer);
-        return(0);
+                  const unsigned char* sig, const unsigned char* pubKey)
+{
+    if (!sig || !msg)
+    {
+        return 0;
     }
 
-    // Assemble signed message in one array.
-    memcpy(signedmsg, msg, msglen);
-    memcpy(signedmsg + msglen, sig, crypto_sign_BYTES);
-
-    unsigned long long bufferlen = 0;
-    int result = crypto_sign_open(msgbuffer, &bufferlen,
-                                  signedmsg, msglen + crypto_sign_BYTES, pubKey);
-    free(msgbuffer);
-    free(signedmsg);
-    if (result) {
-        return(1);
-    } else {
-        // crypto_sign_open() returns -1 on failure.
-        return(0);
-    }
+    int result = crypto_sign_verify_detached(sig, msg, msglen, pubKey);
+
+    return (result == 0) ? 1 : 0;
```

Comment:
This could be simplified just returning `result`, or `!!result` if it's needed to return `0` or `1`.